cluster/rpc/server.go

```diff
@@ -179,3 +184,83 @@ func toRPCError(err error) error {
 	}
 	return status.Error(ec, err.Error())
 }
+
+// QuerierStream is experimental. QuerierStream is triggered when a querier node connects
+// to this metadata node via this bidirectional gRPC stream. We register the querier node
+// with the querier manager and send class tenant data updates to the querier node when
+// they appear on this querier's class tenant data updates channel.
+// We don't currently expect any messages from the querier node, but just use the existence
+// of the stream to keep the connection alive so we can send events to the querier node.
+// This function blocks until the stream is closed by the querier node.
+func (s *Server) QuerierStream(stream cmd.ClusterService_QuerierStreamServer) error {
+	// set up a querier and register it
+	q := querier.NewQuerier()
+	s.querierManager.Register(q)
+	defer s.querierManager.Unregister(q)
+
+	// streamClosed should receive a message when the stream is closed by the querier node,
+	// so we can close the goroutine that sends class tenant data updates.
+	streamClosed := make(chan struct{})
+	// returnErr should be set if there is an error in the goroutines below that should
+	// be returned to the caller.
+	var returnErr error
+
+	wg := sync.WaitGroup{}
+	wg.Add(2)
+
+	// Start a goroutine to wait until the stream is closed by the querier node.
+	// If we extend the protocol to send messages from the querier to the metadata node,
+	// then they'll be handled in this goroutine.
+	enterrors.GoWrapper(func() {
+		defer wg.Done()
+		defer func() { streamClosed <- struct{}{} }()
+
+		// Wait here until the stream is done (for now, we aren't expecting any messages from the querier)
+		_, err := stream.Recv()
+		if err == io.EOF {
+			// this error is expected when the stream is closed by the querier
+			return
+		}
+		if err != nil {
+			// unexpected error
+			returnErr = fmt.Errorf("querier register stream recv: %w", err)
+			return
+		}
+	}, s.log)
+
+	// This channel will receive a message when a class/tenant's data has been updated
+	classTenantDataUpdates := q.ClassTenantDataUpdates()
+
+	// Start a goroutine to send class tenant data updates to the querier node.
+	//
+	enterrors.GoWrapper(func() {
+		defer wg.Done()
+		for {
+			// TODO think more about this select and stream context, do we need streamClosed?
```

Comment:
It would be nice if that PR can include the select statement exiting on the context from the stream being closed. We forget in too many places to handle context and i'd rather we avoid adding new ones (even with a TODO).